### PR TITLE
feat: add ability to send/receive message attributes

### PR DESCRIPTION
**Pull request by wilsonjord**
## Proposed Changes

Resolves #.

Changes proposed in this pull request:

- add/receive message attributes with message

## Production Changes

The following production changes are required to deploy these changes:

- None

## Review

Check the box that applies to this code review.  If necessary please seek help with adding a checklist guide for the reviewer.
When assigning the code review please consider the expertise needed to review the changes.

- [ ] This is a content (documentation, web page etc) only change.
- [ ] This is a minor change (meta data, bug fix, improve test coverage etc).
- [X] This is a larger change (new feature, significant refactoring etc).  Please use the code review guidelines to add a checklist below to guide the code reviewer.

### CODE REVIEW

File: aws/sqs/sqs.go [hunk 1]
```diff
@@ -157,10 +169,11 @@ func (s *SQS) receiveMessage(ctx context.Context, input *sqs.ReceiveMessageInput
...
 	input := sqs.ReceiveMessageInput{
-		QueueUrl:            aws.String(queueURL),
-		MaxNumberOfMessages: 1,
-		VisibilityTimeout:   visibilityTimeout,
-		WaitTimeSeconds:     20,
+		QueueUrl:              aws.String(queueURL),
+		MaxNumberOfMessages:   1,
+		VisibilityTimeout:     visibilityTimeout,
+		WaitTimeSeconds:       20,
+		MessageAttributeNames: []string{"All"},
```
Comment (junghao) [inline aws/sqs/sqs.go:176]: If everything works as usual without this `MessageAttributeNames` then I'd suggest we don't add it.

File: aws/sqs/sqs.go [hunk 2]
```diff
@@ -179,11 +192,27 @@ func (s *SQS) Delete(queueURL, receiptHandle string) error {
...
 	params := sqs.SendMessageInput{
 		QueueUrl:    aws.String(queueURL),
 		MessageBody: aws.String(body),
 	}
 
+	// add message attributes
+	if len(attributes) > 0 {
+		params.MessageAttributes = map[string]types.MessageAttributeValue{}
+		for k, v := range attributes {
+			params.MessageAttributes[k] = types.MessageAttributeValue{
```
Comment (junghao) [inline aws/sqs/sqs.go:209]: Same suggestion: https://github.com/aws/aws-sdk-go/blob/main/service/sqs/api.go#L7501

**Review by junghao — COMMENTED** [review]
Some discussions.

File: aws/sqs/sqs.go [hunk 3]
```diff
@@ -139,10 +141,20 @@ func (s *SQS) receiveMessage(ctx context.Context, input *sqs.ReceiveMessageInput
 		case len(r.Messages) == 1:
 			raw := r.Messages[0]
 
+			msgAttributes := map[string]string{}
+			for k, v := range raw.MessageAttributes {
+				if aws.ToString(v.DataType) == "Binary" {
```
[the remark this answers is not in the document]
Comment (wilsonjord) [reply aws/sqs/sqs.go:146]: Thanks Howard...good point here, and thanks for the link! I will update the PR to let the caller decide what to do with the result.

**Comment by wilsonjord** [comment]
I no longer have an immediate usecase for message attributes, therefore I'm closing this; thank you for your time regardless.

**Pull request closed without merging**